TradeItIosEmsApi/TradeItConnector.m

```diff
@@ -397,65 +397,64 @@ - (void)oAuthDeleteLink:(TradeItLinkedLogin *)linkedLogin {
-                NSLog(@"ERROR from EMS server response=%@ error=%@", response, error);
-                TradeItErrorResult *errorResult = [TradeItErrorResult errorWithSystemMessage:@"error sending request to ems server"];
-                dispatch_async(dispatch_get_main_queue(), ^(void) {
-                    completionBlock(errorResult, nil);
-                });
-
-                return;
-            }
-
-            NSMutableString *jsonResponse = [[NSMutableString alloc] initWithData:data encoding:NSUTF8StringEncoding];
-
-//             NSLog(@"----------Response %@----------", [[request URL] absoluteString]);
-//             NSLog(jsonResponse);
-
-            //first convert to a generic result to check the type
-            TradeItResult *tradeItResult = [TradeItRequestResultFactory buildResult:[TradeItResult alloc]
-                                                                         jsonString:jsonResponse];
-
-            if ([tradeItResult.status isEqual:@"ERROR"]) {
-                TradeItErrorResult * errorResult;
-
-                if (![tradeItResult isKindOfClass:[TradeItErrorResult class]]) {
-                    errorResult = (TradeItErrorResult *)[TradeItRequestResultFactory buildResult:[TradeItErrorResult alloc]
-                                                                                      jsonString:jsonResponse];
-                } else {
-                    errorResult = (TradeItErrorResult *) tradeItResult; //this type of error caused by something wrong parsing the response
-                }
-                tradeItResult = errorResult;
-            }
-            dispatch_async(dispatch_get_main_queue(),^(void){completionBlock(tradeItResult, jsonResponse);});
-        }] resume];
+        [[session dataTaskWithRequest:request completionHandler:^(NSData * _Nullable data, NSURLResponse * _Nullable response, NSError * _Nullable error) {
+              NSHTTPURLResponse *httpResponse = (NSHTTPURLResponse *)response;
+              if ((data == nil) || ([httpResponse statusCode] != 200)) {
+                  //error occured
+                  NSLog(@"ERROR from EMS server response=%@ error=%@", response, error);
+                  TradeItErrorResult *errorResult = [TradeItErrorResult errorWithSystemMessage:@"error sending request to ems server"];
+                  dispatch_async(dispatch_get_main_queue(), ^(void) {
+                      completionBlock(errorResult, nil);
+                  });
+                  return;
+              }
+
+              NSMutableString *jsonResponse = [[NSMutableString alloc] initWithData:data encoding:NSUTF8StringEncoding];
+
+              TradeItResult *result = [TradeItRequestResultFactory buildResult:[ResultClass alloc] jsonString:jsonResponse];
+
+              if ([result.status isEqual:@"ERROR"]) {
+                  result = [TradeItRequestResultFactory buildResult:[TradeItErrorResult alloc] jsonString:jsonResponse];
+              }
```

Comment:
There are 2 types of errors that can happen here:

1. Error from EMS: the `result` will be a `TradeItErrorResult` because the `ResultClass` could not be constructed. It will enter this `if` statement and build the `TradeItErrorResult` properly.
2. Error parsing JSON: the `result` will be a `TradeItErrorResult` for invalid JSON. It will get constructed again, but returns the same error.